### ISSUE TITLE
`process.argv` is changed

### ISSUE THREAD

**Issue by coolwanglu** [atom/atom-shell#1248]
```javascript
//main.js
console.log(process.argv);
```

When I run the app with `atom-shell . a -b c`, I got the output `atom-shell -b . a c`
I think that atom-shell does this to extract the arguments useful for atom-shell, but my app also need the original order of arguments. Is this a bug, or there are other APIs for this?

**Comment by deepak1556** [comment]
not exactly atom-shell specific, just happen to be the cause of using `base::commandLine` https://github.com/atom/atom-shell/blob/e98953a5a563648a69b6cb8535d1f27febeec887/atom/common/node_bindings.cc#L173 which follows the pattern https://code.google.com/p/chromium/codesearch#chromium/src/base/command_line.h&sq=package:chromium&l=217 . This just propagated to the environment creation for the process object.

**Comment by bwin** [comment]
If you prepend `--` like `atom-shell -- . a -b c` the order is preserved.
But you got an additiional arg `--` (which is pretty easy to discard).

**Comment by bwin** [comment]
If you depend on user-provided cli args, this is a serious problem.
It groups all `--`-args, then all `-`-args and then the rest.
If your args are in the format `--target 0.1.2` this f*cks things up. You're fine, if you use args like `--target=0.1.2`. (Or if you can prepend `--` before the actual args.)

**Comment by bwin** [comment]
Nonetheless this should probably be fixed...

**Comment by maxogden** [comment]
just ran into this bug. for example, when I call my CLI atom-shell app with these args:

```
find-data http://irsa.ipac.caltech.edu/ibe/data/wise/merge/merge_p1bm_frm/ --workers=2
```

I would expect:

```
[ '/usr/local/lib/node_modules/atom-shell/dist/Atom.app/Contents/MacOS/Atom',
  '/usr/local/bin/find-data',
  'http://foo.com',
  '--workers=2']
```

but instead I get

```
[ '/usr/local/lib/node_modules/atom-shell/dist/Atom.app/Contents/MacOS/Atom',
  '--workers=2',
  '/usr/local/bin/find-data',
  'http://foo.com' ]
```

... which breaks many CLI arg parsing tools from NPM :(

**Comment by bwin** [comment]
@maxogden As long as your args are like `--workers=2` and not `--workers 2` I think it should work with arg-parsers like `commander` or `minimist` etc. since the order of the unnamed<sup>1</sup> args seems to be preserved.
But, I could be overlooking something here.

(<sup>1</sup> like argv._ in `minimist`: all the arguments that didn't have an option associated with them)

As a work-around you could also prepend `--` to preserve the order of all args.

**Comment by maxogden** [comment]
@bwin yea its an executable in my PATH that looks like this:

```js
#!/usr/bin/env atom-shell
var args = require('minimist')(process.argv.slice(2))
// atom-shell code goes here
```

minimist would work if the `--` arg got moved to the end, but since it gets moved to index 1 it breaks the parsing.

**Comment by bwin** [comment]
@maxogden What about using `./atom find-data -- http://foo.com --workers=2` *and* parsing `process.argv.slice(3)`? It's ugly-as-f*ck, but should work, right? I'll give it a try...

**Comment by bwin** [comment]
Seems right. `Minimist` gives me `{ _: [ 'http://foo.com' ], workers: 2 }` in that case.

**Comment by maxogden** [comment]
@bwin yes there are definitely workarounds (env vars etc), but IMO atom-shell should match the behavior of node/iojs

**Comment by bwin** [comment]
Yeah, as I said, I think it should be fixed. I was just trying to find a *decent* work-around for your use-case.
BTW your profile pic :+1:

**Comment by maxogden** [comment]
@bwin thanks for the help! and the pic is taken at GitHub's office in SF, you should get a pic there if you ever visit

**Comment by bwin** [comment]
No way! How cool is that! It seems like the guys at GitHub really rock.
(Sidenote: as a stupid German I presumed it was the real White House... :hurtrealbad:)

**Comment by zcbenz** [comment]
We should probably save `argv` somewhere instead of relying on `base::CommandLine` to get it.

**Comment by bwin** [comment]
:+1:

**Comment by zcbenz** [comment]
Fixed by #1256.
